funflow-checkpoints/src/Control/FunFlow/Checkpoints.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NoImplicitPrelude     #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeOperators         #-}
+module Control.FunFlow.Checkpoints
+  ( CheckpointException(..)
+  , CheckpointT
+  , checkpoint
+  , runToCompletion
+  , runToCheckpoint
+  , runToCheckpointEither
+  ) where
+
+import           Control.Arrow
+import           Control.Arrow.Free
+import           Control.Category
+import           Control.Exception.Base (Exception, throw)
+import           Control.FunFlow
+import           Data.Either            (Either (..))
+import           Data.Eq                (Eq, (==))
+import           Data.Function          (flip, ($))
+import           Data.Maybe             (Maybe (..))
+import           Data.Proxy
+import           Data.String            (String)
+import           Data.Tuple             (fst, snd)
+import           Data.Typeable
+import           Text.Show              (Show)
+
+data CheckpointException =
+    -- | A checkpoint has been defined multiple times in the flow
+    RepeatedCheckpoint String
+    -- | A conditional branch has a checkpoint in.
+  | CheckpointInConditional String
+    -- | A branch responsible for catching errors has a checkpoint in.
+  | CheckpointInCatch String
+    -- | No matching checkpoint has been defined in the flow
+  | NoCheckpoint String
+  deriving (Eq, Show)
+
+instance Exception CheckpointException
+
+data CheckpointRef a = CheckpointRef (Proxy a) String
+  deriving Eq
+
+-- | Get the name of a constructed checkpoint
+checkpointName :: CheckpointRef a -> String
+checkpointName (CheckpointRef _ x) = x
+
+-- | A checkpoint arrow transfomer.
+data CheckpointT arr (eff :: * -> * -> *) ex a b where
+  Checkpoint :: Typeable a => CheckpointRef a -> CheckpointT arr eff ex a a
+  Unchecked :: arr eff ex a b -> CheckpointT arr eff ex a b
+  Seq :: CheckpointT arr eff ex a b -> CheckpointT arr eff ex b c -> CheckpointT arr eff ex a c
+  Par :: CheckpointT arr eff ex a b -> CheckpointT arr eff ex c d -> CheckpointT arr eff ex (a,c) (b,d)
+  Fanin :: CheckpointT arr eff ex a c -> CheckpointT arr eff ex b c -> CheckpointT arr eff ex (Either a b) c
+  Catch :: CheckpointT arr eff ex a b -> CheckpointT arr eff ex (a,ex) b -> CheckpointT arr eff ex a b
+
+instance Category (arr eff ex) => Category (CheckpointT arr eff ex) where
+  id = Unchecked id
+  (.) = flip Seq
+instance Arrow (arr eff ex) => Arrow (CheckpointT arr eff ex) where
+  arr f = Unchecked $ arr f
+  first f = Par f id
+  second = Par id
+  (***) = Par
+instance ArrowChoice (arr eff ex) => ArrowChoice (CheckpointT arr eff ex ) where
+  f +++ g = (f >>> arr Left) ||| (g >>> arr Right)
+  f ||| g = Fanin f g
+instance ArrowError ex (arr eff ex) => ArrowError ex (CheckpointT arr eff ex) where
+  f `catch` g = Catch f g
+
+instance ArrowFlow eff ex (arr eff ex) => ArrowFlow eff ex (CheckpointT arr eff ex) where
```

Comment:
Note: if you export a constructor for `Unchecked` (like `lift` for monad transformers). Then you can remove the dependency in the funflow package, and define this instance in funflow instead.

Reply:
I think I'd rather have the dependency this way around, since what with the requirements on `eff, ex` etc, `CheckpointT` isn't really suitable for use in other places, whereas I'd rather not have funflow depending on this.

Reply:
Honestly, I'm not convinced that you actually need to have explicit `eff` and `ex`: you don't use them anywhere, do you?

Maybe you can just drop these? (independently of whether you want the dependency to be this or the other way around)